_posts/2023-04-10-opensource-perf.markdown

```diff
@@ -0,0 +1,127 @@
+
+Apache James relies on OpenSearch as its search engine for its email database. It takes as a dependency the [opensearch-java client](https://github.com/opensearch-project/opensearch-java). We frequently run performance tests with custom [Gatling](https://gatling.io/) benchmarks. We specifically wanted to regression test performance when we migrated from ElasticSearch 7.10 to OpenSearch for licenses reasons (as an Apache project, we must use licenses compatible with the Apache License 2.0). We realized the OpenSearch requests were slower and started investigating why.
+
+### Identifying the performance issue
+
+Flamegraphs are powerful visualization tools that can help developers analyze the execution flow of their code and identify bottlenecks. By analyzing the flamegraphs, I was able to identify specific areas of code that were causing the performance issue. We used async-profiler to generate a flamegraph and found that frequent SPI lookups were causing the performance issue.
+
+[opensearch_2.4.0.zip](https://github.com/opensearch-project/opensearch-java/files/10334079/opensearch_2.4.0.zip) is the flame graph we diagnosed this issue from. It was taken with <https://github.com/jvm-profiling-tools/async-profiler>  from within Apache James container):
+
+`./profiler -d 120 -e itimer -f opensearch_2.4.0.html 1`
+
+
+<img src="/assets/media/blog-images/2023-04-10-opensource-perf/flame1.png" alt="Flame graph general overview"/>{: .img-fluid }
+
+
+Then searching for opensearch using the search widget in the top left corner of the flame graph:
+
+<img src="/assets/media/blog-images/2023-04-10-opensource-perf/flame2.png" alt="Flame graph: OpenSearch driver threads"/>{: .img-fluid }
+
+-   The driver thread (on the right) with the httpclient event loop
+-   Apache James calls to the OpenSearch driver (small stacktraces in the center)
+
+Observations on the driver event loop:
+
+-   We spend a huge amount of resources doing SPI calls in order to locate the JSON parser implementation. It not only eats CPU/heap allocations but it also blocks the HTTP event loop, which can be catastrophic (the James app is not OpenSearch heavy so that is not a big issue to me).
+-   Unsurprisingly JSON parsing takes 11% of the driver's CPU and 24% of heap allocation.
+-   Event loop busyness (pushing stuff to the kernel network stack, SSL) takes around 60% of the event loop CPU, 25% of heap allocations, which again feels normal to me.
+-   2.3% of the event loop CPU is our binding to Reactor reactive library, which, again is normal: converting futures and enqueuing tasks takes time.
+
+To mitigate calls to the SPI, we submitted a [small change](https://github.com/opensearch-project/opensearch-java/pull/293/files) to JsonValueParser.java that addressed the issue. This led to a 50x speedup of the JSON parsing by getting rid of the SPI lookups, not even taking into account the blocking operations... Huge win!
```

Comment:
small change -> a change

remove /files from the PR URL

link JsonValueParser.java